### PR TITLE
Adding Expression Metrics & Metric Types to Metric Page

### LINKED ISSUE #1638: Expression (?) metrics
### Contributions

- [x] I have read the contribution docs, and understand what's expected of me.

### Link to the page on docs.getdbt.com requiring updates

https://docs.getdbt.com/docs/building-a-dbt-project/metrics

### What part(s) of the page would you like to see updated?

Fantastic expression/derived/composite metrics, and how to define them :)

### Additional information

https://github.com/dbt-labs/dbt-core/pull/5027

### PR DESCRIPTION
**Pull request by callum-mcdata**
## Description & motivation
This PR adds information to the `metrics.md` file about expression metrics. Additionally, it adds a table of supported metric types. NOTE: These metric type definitions actually live within the metrics package but I believe it is important for that information to live in the docs and not just the package README. 

This PR satisfies Issue #1638 

## Prerelease docs
If this change is related to functionality in a prerelease version of dbt (delete if not applicable):
- [X] I've added versioning components, as described in ["Versioning Docs"](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/versioningdocs.md)
- [ ] I've added a note to the prerelease version's [Migration Guide](https://github.com/dbt-labs/docs.getdbt.com/tree/current/website/docs/guides/migration/versions)

### CODE REVIEW

**Comment by jasnonaz** [comment]
This is so awesome thanks for putting together @callum-mcdata!! Looks like the netlify build failed - any thoughts on what we might need to tweak?

**Comment by callum-mcdata** [comment]
Hmm if I had to hazard a guess I would say this is probably because of some `<VersionBlock>` shenanigans. I wonder if `1.2` hadn't been defined as a version block and is failing because of it. Let me see if I can peer behind the netlify curtain

**Comment by callum-mcdata** [comment]
Well when a certain someone hasn't done a pull from `current` since he started, no wonder there are some issues 😅 . Just pulled in the most recent changes and I believe that should fix the issue

**Review by jasnonaz — APPROVED** [review]
This is fantastic work! Thanks for putting this together and powering through docusaurus issues. Non blocking question for you: Do you think we should have info about the metrics package in here / link to the metrics package. Feels like this is important information for folks who are getting into metrics.

**Comment by callum-mcdata** [comment]
@jasnonaz that is a really good point - definitely should have some links to the metrics package, just as the metrics package should have some links to this doc so people can cross reference. Let me add an additional section linking to the package

File: website/docs/docs/building-a-dbt-project/metrics.md
Comment (jtcohen6) [inline website/docs/docs/building-a-dbt-project/metrics.md:30]: [Non-blocking] Thoughts on updating this language? It still refers to:
- Metrics being brand-new in v1.0 — less true now than it was last December
- Metrics being **experimental** / liable to change in new minor versions — this is still true, but we anticipate removing this "experimental" designation in a few months!

Happy for those changes to come in a subsequent PR, since it's adjacent scope

File: website/docs/docs/building-a-dbt-project/metrics.md
Comment (jtcohen6) [inline website/docs/docs/building-a-dbt-project/metrics.md]: 👍  Definitely think it's more than appropriate to prominently link to the `dbt-metrics` package from this page

[Blocking] I think I know what you're saying here, but the wording is a bit tricky. Do you mean:
- "For more information on querying metrics that are defined in your dbt project"
- "For more information on how metrics queries are defined"
?

File: website/docs/docs/building-a-dbt-project/metrics.md
Comment (jtcohen6) [inline website/docs/docs/building-a-dbt-project/metrics.md]: [Blocking] From a quick visual inspection, I think this is the only change in the table between v1.0/v1.1 &rarr; v1.2:
- `model` + `Required?`: `yes (no for expression metrics)`, instead of `yes` (= always) 

Is that right?

I think our `VersionBlock` logic might be fancy enough to make this work:
```suggestion
| model       | The dbt model that powers this metric                       | dim_customers                   | <VersionBlock firstVersion="1.2">yes (no for `expression` metrics)</VersionBlock><VersionBlock lastVersion="1.1">yes</VersionBlock> |
```
I tried that locally and it seemed to do the trick. Then we don't need to duplicate the full table contents for different docs versions.

File: website/docs/docs/building-a-dbt-project/metrics.md
Comment (jtcohen6) [inline website/docs/docs/building-a-dbt-project/metrics.md:132]: [Non-blocking]

> As long as the two+ base metrics (the metrics that comprise the `expression` metric) share the specified `time_grains` and `dimensions`

Ah - do we have validation for this in `dbt-core`? In the `dbt-metrics` package?

File: website/docs/docs/building-a-dbt-project/metrics.md
Comment (jtcohen6) [inline website/docs/docs/building-a-dbt-project/metrics.md:153]: [Non-blocking] Is this the syntax we'd recommend, over syntax like:
```yaml
    sql: "{{ metric('total_revenue') ~ "/" ~ metric('count_of_customers') }}"
```
(Does that alt syntax even work?)

**Review by jtcohen6 — CHANGES_REQUESTED** [review]
@callum-mcdata Thanks for taking this on! I left a few comments, and called out which ones are [blocking] or [non-blocking] for merging this PR.

Could you also add a note to [the "New and changed documentation" section of the v1.2 migration guide](https://github.com/dbt-labs/docs.getdbt.com/blob/current/website/docs/guides/migration/versions/upgrading-to-v1.2.md), describing the change at a high level (one sentence) and linking to this page?

File: website/docs/docs/building-a-dbt-project/metrics.md
Comment (callum-mcdata) [reply website/docs/docs/building-a-dbt-project/metrics.md]: Wow, yea the language here is ... not great 😅 . I will update to:

`For more information on querying the metrics defined in your dbt project`

File: website/docs/docs/building-a-dbt-project/metrics.md
Comment (callum-mcdata) [reply website/docs/docs/building-a-dbt-project/metrics.md:132]: We have "validation" for this in the `dbt_metrics` package. The macro constructs a list of acceptable dimensions from all metrics used and then the provided dimensions are compared against that list. The macro errors out if it is an invalid dimension.

File: website/docs/docs/building-a-dbt-project/metrics.md
Comment (callum-mcdata) [reply website/docs/docs/building-a-dbt-project/metrics.md:153]: I tested the alt syntax and it didn't appear to work. Even testing a few different options around quotation placement returned unicode errors. 

To me that is fine anyway because I wouldn't want users to have to worry about concatenation within the sql field. Self contained makes most sense to me and personally feels the most ergonomic.

File: website/docs/docs/building-a-dbt-project/metrics.md
Comment (callum-mcdata) [reply website/docs/docs/building-a-dbt-project/metrics.md]: Resolved!

File: website/docs/docs/building-a-dbt-project/metrics.md
Comment (callum-mcdata) [reply website/docs/docs/building-a-dbt-project/metrics.md:30]: I agree - opened #1690 to remember this suggestion and track it in a future PR!

File: website/docs/docs/building-a-dbt-project/metrics.md
Comment (jtcohen6) [inline website/docs/docs/building-a-dbt-project/metrics.md:100]: Idea in making this change is that we can have just one table of "Available properties," rather than duplicating it here + below (beneath `<VersionBlock lastVersion="1.1">`)

**Comment by callum-mcdata** [comment]
@jtcohen6 thanks for all the helpful comments! I've resolved the blocking issues and addressed the non-blocking issues (opened an issue for one of them).  The upgrading page has also been updated with the following line:

- Updated `metrics.md` to reflect the addition of the `metric` function and support for `expression` metrics.

Let me know if there is anything blocking for this PR!

File: website/docs/docs/building-a-dbt-project/metrics.md
Comment (callum-mcdata) [reply website/docs/docs/building-a-dbt-project/metrics.md:100]: Accepted changes!

File: website/docs/guides/migration/versions/upgrading-to-v1.2.md
Comment (jtcohen6) [inline website/docs/guides/migration/versions/upgrading-to-v1.2.md]: Two things:
- The intent here is not to keep a pure changelog, but rather to give users a "highlight reel" of what's changed, and links to read more
- You'll need to resolve a merge conflict before merging (most PRs for v1.2 target updates to this page)
```suggestion
- [**Metrics**](building-a-dbt-project/metrics) now support an `expression` type (metrics-on-metrics), as well as a `metric()` function to use when referencing metrics from within models or `expression`-type metrics.
```

File: website/docs/docs/building-a-dbt-project/metrics.md
Comment (jtcohen6) [reply website/docs/docs/building-a-dbt-project/metrics.md:100]: Not quite. Here's a commit with what I mean: https://github.com/dbt-labs/docs.getdbt.com/commit/54b9b818db0121d2e8fea92690d169933585cdaf

File: website/docs/docs/building-a-dbt-project/metrics.md
Comment (jtcohen6) [inline website/docs/docs/building-a-dbt-project/metrics.md:112]: Any reason why we can't show this table (or at least most of it) to users of v1.0 + v1.1 as well? I'd propose removing the `VersionBlock` wrapping around the whole table, and then putting something like this for the last line:
```
| expression | <VersionBlock firstVersion="1.2">This metric type is defined as any **non-aggregating** calculation of 1 or more metrics</VersionBlock><VersionBlock lastVersion="1.1">Not yet available — added in v1.2</VersionBlock> |
```

**Review by jtcohen6 — APPROVED** [review]
LGTM pending these two changes!
- [X] Update migration guide note (to resolve merge conflict)
- [X] DRY up "Available properties" table

**CM Edit**: *Tasks accomplished on 7/13/22*

File: website/docs/docs/building-a-dbt-project/metrics.md
Comment (callum-mcdata) [reply website/docs/docs/building-a-dbt-project/metrics.md:112]: I'm on board with this idea - I think it was something missing from the pre-1.2 versions anyway so if I can backfill it that's a bonus in my book.

**Comment by callum-mcdata** [comment]
@jtcohen6 The issues raised above have been resolved. For some reason though, I swear I read that one of the comments was around updating the **discussions** section but I'm unable to find that now. Was I dreaming and imagining that or has it just hidden itself in this comment thread?

File: website/docs/docs/building-a-dbt-project/metrics.md
Comment (jtcohen6) [reply website/docs/docs/building-a-dbt-project/metrics.md:100]: @callum-mcdata Still need to delete the duplicated version of this table below as well:

Screenshot 2022-07-14 at 01 50 40

Comment (callum-mcdata) [reply website/docs/docs/building-a-dbt-project/metrics.md:100]: Odd - when I spun it up on my local there was only one table 🤔  I wonder if I messed up on the merge conflict from current. Let me fix now!

Comment (callum-mcdata) [reply website/docs/docs/building-a-dbt-project/metrics.md:100]: Yep that was it! Fixed it:

image

**Review by jtcohen6 — APPROVED** [review]
Let's ship it!

**Comment by callum-mcdata** [comment]
@annafil @runleonarun I'll let y'all press the big green button as you are the codeowners for this repo. Or if you'd prefer me to press it, let me know!

**Comment by jtcohen6** [comment]
@callum-mcdata I think you're good to merge yourself, since you did the leg work here! But I'll press it if you don't want to ;)

**Comment by callum-mcdata** [comment]
GONNA PRESS THAT BUTTON THEN! 🕺

**Pull request merged**
